From 48d37eceb333ab38a98d3697c6e3f55dd2535ba5 Mon Sep 17 00:00:00 2001
From: Zhao Wei Liew <zhaoweiliew@gmail.com>
Date: Fri, 2 Nov 2018 00:33:14 +0000
Subject: [PATCH 17/17] Cache SOC ID checks for future queries

Currently, get_soc_id() is queried on every SOC ID check in the
SDM660, MSM8916, and MSM8974-family HALs. This results in
extraneous file operations on every SOC ID check.

Cache the result of get_soc_id() during the first query to
reduce the number of file operations being made.

This also brings back the behaviour in LineageOS 15.1 when the
HALs were still stored in device/qcom/common.

Change-Id: Ic17dbf12e7f9ecdb47b73a580f467df9ad630aa0
---
 power-660.c  | 13 ++++++-------
 power-8916.c | 13 ++++++-------
 power-8974.c | 13 ++++++-------
 3 files changed, 18 insertions(+), 21 deletions(-)

diff --git a/power-660.c b/power-660.c
index f9f60e1..9b1d666 100644
--- a/power-660.c
+++ b/power-660.c
@@ -142,19 +142,18 @@ static int set_power_profile(int profile)
 }
 
 /**
- * If target is SDM630:
- *     return true
- * else:
- *     return false
+ * Returns true if the target is SDM630.
  */
 static bool is_target_SDM630(void)
 {
-    static bool is_SDM630 = false;
+    static int is_SDM630 = -1;
     int soc_id;
 
+    if (is_SDM630 >= 0)
+        return is_SDM630;
+
     soc_id = get_soc_id();
-    if (soc_id == 318 || soc_id == 327)
-        is_SDM630 = true;
+    is_SDM630 = soc_id == 318 || soc_id == 327;
 
     return is_SDM630;
 }
diff --git a/power-8916.c b/power-8916.c
index 190b336..66b57e7 100644
--- a/power-8916.c
+++ b/power-8916.c
@@ -60,19 +60,18 @@ const char *scaling_min_freq[4] = {
 };
 
 /**
- * If target is 8916:
- *     return true
- * else:
- *     return false
+ * Returns true if the target is MSM8916.
  */
 static bool is_target_8916(void)
 {
-    static bool is_8916 = false;
+    static int is_8916 = -1;
     int soc_id;
 
+    if (is_8916 >= 0)
+        return is_8916;
+
     soc_id = get_soc_id();
-    if (soc_id == 206 || (soc_id >= 247 && soc_id <= 250))
-        is_8916 = true;
+    is_8916 = soc_id == 206 || (soc_id >= 247 && soc_id <= 250);
 
     return is_8916;
 }
diff --git a/power-8974.c b/power-8974.c
index adad79e..9c2efa9 100644
--- a/power-8974.c
+++ b/power-8974.c
@@ -53,19 +53,18 @@
 static int first_display_off_hint;
 
 /**
- * If target is 8974pro:
- *     return true
- * else:
- *     return false
+ * Returns true if the target is MSM8974AB or MSM8974AC.
  */
 static bool is_target_8974pro(void)
 {
-    static bool is_8974pro = false;
+    static int is_8974pro = -1;
     int soc_id;
 
+    if (is_8974pro >= 0)
+	    return is_8974pro;
+
     soc_id = get_soc_id();
-    if (soc_id == 194 || (soc_id >= 208 && soc_id <= 218))
-        is_8974pro = true;
+    is_8974pro = soc_id == 194 || (soc_id >= 208 && soc_id <= 218);
 
     return is_8974pro;
 }
-- 
2.17.1

